### PR TITLE
Ade20k benchmark

### PR DESCRIPTION
**Pull request by A-Jacobson**
ade20k benchmark in the style of the resnet benchmark. definitely not generated by gpt.

Notes:

- Added `download_ade20k.py` - considered adding a `download=True` flag to the main script but it felt complicated to include along with the streaming object. Thought this may be a more simple compromise but am open to suggestions.
- mcloud run config still points to my fork. will need to be changed once this is merged.
- light on tests.. probably?

Additionally:

Did some renaming and refactoring to keep this in line with the resnet blog. Examples include:

1. local dataset `splits` arg from `training`, `validation`, `test` - > `train`, `val`, `test`
2. `build_ade20k_dataloader` -> `build_ade20k_dataspec` - this is was the resnet example is doing and these functions are actually retuning dataspecs
3. `composer_deeplab` -> `build_composer_deeplab` - now takes pytorch initialization functions rather than composer `Initializer` objects

### CODE REVIEW

File: deeplab/data.py [hunk 1]
```diff
@@ -0,0 +1,684 @@
...
+                            **dataloader_kwargs)
+    device_transform_fn = NormalizationFn(mean=IMAGENET_CHANNEL_MEAN,
+                                          std=IMAGENET_CHANNEL_STD,
+                                          ignore_background=ignore_background)
+
+    return DataSpec(dataloader=dataloader,
+                    device_transforms=device_transform_fn)
+
+
+def build_synthetic_ade20k_dataspec(
```
Comment (Landanjs) [inline deeplab/data.py:232]: Is this used anywhere? If not, should we delete it? I know synthetic datasets can be useful to measure model throughput, but I don't think we are setting this up in the benchmarks (yet).

File: deeplab/README.md [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+        <img alt="Chat @ Slack" src="https://img.shields.io/badge/slack-chat-2eb67d.svg?logo=slack">
+    </a>
+    <a href="https://github.com/mosaicml/benchmarks/blob/main/LICENSE">
+        <img alt="License" src="https://img.shields.io/badge/License-Apache%202.0-green.svg?logo=slack">
+    </a>
+</p>
+<br />
+
+# Mosaic DeepLabV3
+This folder contains starter code for training [mmsegmentation DeepLabV3 architectures](https://github.com/open-mmlab/mmsegmentation/tree/master/configs/deeplabv3plus) using our most efficient training recipes (see our [benchmark blog](https://www.mosaicml.com/blog/behind-the-scenes) or [recipies blog](https://www.mosaicml.com/blog/mosaic-image-segmentation) for details). These recipes were developed to hit baseline accuracy on [ADE20K](https://groups.csail.mit.edu/vision/datasets/ADE20K/) 5x faster or to maximize ADE20K accuracy over long training durations. Although these recipes were developed for training DeepLabV3 on ADE20k, they could be used to train other segmentation models on other datasets. Give it a try!
```
Comment (Landanjs) [inline deeplab/README.md:26]:
```suggestion
This folder contains starter code for training [mmsegmentation DeepLabV3+ architectures](https://github.com/open-mmlab/mmsegmentation/tree/master/configs/deeplabv3plus) using our most efficient training recipes (see our [benchmark blog post](https://www.mosaicml.com/blog/behind-the-scenes) or [recipies blog post](https://www.mosaicml.com/blog/mosaic-image-segmentation) for details). These recipes were developed to hit baseline mean Intersection-over-Union (mIoU) on [ADE20K](https://groups.csail.mit.edu/vision/datasets/ADE20K/) 5x faster or to maximize ADE20K mIoU over long training durations. Although these recipes were developed for training DeepLabV3+ on ADE20k, they could be used to train other segmentation models on other datasets. Give it a try!
```

Comment (Landanjs) [reply deeplab/README.md:26]: Is accuracy -> mIoU pedantic? 😅

File: deeplab/README.md [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+
+The specific files in this folder are:
+* `model.py` - A function to create a [ComposerModel](https://docs.mosaicml.com/en/v0.11.0/composer_model.html) from an mmsegmentation DeepLabV3 model
+* `data.py` - A [MosaicML streaming dataset](https://docs.mosaicml.com/projects/streaming/en/latest/) for ADE20K and a PyTorch dataset for a local copy of ADE20K
+* `download_ade20k.py` - A helper script for downloading ADE20K locally
+* `main.py` - The training script that builds a [Composer](https://github.com/mosaicml/composer) Trainer based on a configuration specified by a yaml
+* `tests/` - A suite of tests to check each training component
+* `yamls/`
+  * `deeplabv3.yaml` - Configuration for a DeepLabV3 training run to be used as the first argument to `main.py`
```
Comment (Landanjs) [inline deeplab/README.md:35]:
```suggestion
  * `deeplabv3.yaml` - Configuration for a DeepLabV3+ training run to be used as the first argument to `main.py`
```

File: deeplab/README.md [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+[epoch=0][batch=17/625]: memory/reserved_mem: 28137488384
+[epoch=0][batch=17/625]: memory/alloc_retries: 3
+[epoch=0][batch=17/625]: trainer/grad_accum: 2
+[epoch=0][batch=17/625]: loss/train/total: 7.1243
+[epoch=0][batch=17/625]: metrics/train/Accuracy: 0.0010
+train          Epoch   0:    3%|▋                        | 17/625 [00:17<07:23,  1.37ba/s, loss/train/total=7.1292]
+```
+# Using Mosaic Recipes
+
+As described in our [Segmentation blog post](https://www.mosaicml.com/blog/mosaic-image-segmentation), we cooked up three recipes to train DeepLabv3 faster and with higher accuracy:
```
Comment (Landanjs) [inline deeplab/README.md:185]:
```suggestion
As described in our [Segmentation blog post](https://www.mosaicml.com/blog/mosaic-image-segmentation), we cooked up three recipes to train DeepLabV3+ faster and with higher accuracy:
```

**Review by Landanjs — COMMENTED** [review]
This was a lot to review haha, great job getting this up!

Do you wandb runs for this?

Requested big changes:
1. Combine `build_ade20k_dataspec` and `build_streaming_ade20k_dataspec`, then add a `is_streaming` argument. This isn't perfect, but it does reduce the data.py and main.py code a bit.
2.  Should the synthetic ADE20k dataset be removed? 
3. I think we should remove the option for v1 streaming since this will be deprecated in the future.
4. Can we remove the `plus` argument for the model and associated code? Reduces the code a bit and I don't have a good reason to keep the code around anymore.
5. I think a smoke test (e.g. `test_trainer` in the ResNet benchmark) is a main component for  benchmarks. Would it be possible to add this? @dblalock for clarification on its importance

Nits:
- Add newline between docstring TL;DR and `Args:`, between `Args:` and `Return:` (mostly in data.py I think?).
- Double check that there is 1 newline at the end of each file

File: deeplab/main.py [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+            import wandb
+        except ImportError as e:
+            raise e
+        if wandb.run:
+            wandb.config.update(OmegaConf.to_container(cfg, resolve=True))
+
+
+def main(config):
+    # update trainer settings based on selected recipe
+    if config.recipe_name:
```
[the remark this answers is not in the document]
Comment (A-Jacobson) [reply deeplab/main.py:45]: we certainly can though there are a lot of arguments without any sort of value checks due to the nature of omegaconf. assuming we don't want to do them all?

File: deeplab/README.md (hunk 2 above)
Comment (A-Jacobson) [reply deeplab/README.md:26]: nope but blog -> blog post is! changed anyway =)

File: deeplab/yamls/deeplabv3.yaml [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+  max_resize_scale: 2.0              # The maximum value the samples can be rescaled
+  final_size: 512                    # The final size of the image and target
+  ignore_background: true            # If true, ignore the background class when calculating the training loss
+  batch_size: 128                    # Training dataloader batch size per device
+
+# Validation Dataset Parameters
+eval_dataset:
+  is_streaming: false                   # Type must be either ['streaming', 'local', 'synthetic']
+  path:  ade20k       # Path to S3 bucket if streaming, otherwise path to local data directory
+  local: /tmp/mds-cache/mds-ade20k/  # Local cache when streaming data
```
[the remark this answers is not in the document]
Comment (A-Jacobson) [reply deeplab/yamls/deeplabv3.yaml]: accidentally left the real paths in there from local testing. will change.

File: deeplab/main.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Landanjs) [reply deeplab/main.py:48]: Fair haha, up to you if you want to include it!

File: deeplab/README.md (hunk 2 above)
Comment (Landanjs) [reply deeplab/README.md]: Hehe, that change was suggested on my PR, so I wanted to do the same :P

File: deeplab/transforms.py [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+
+
+def build_ade20k_transformations(split,
+                                 base_size: int = 512,
+                                 min_resize_scale: float = 0.5,
+                                 max_resize_scale: float = 2.0,
+                                 final_size: int = 512):
+    """Builds the transformations for the ADE20k dataset.
+       Args:
+           split (str): The dataset split to use either 'train', 'val', or 'test'. Default: ``'train```.
```
Comment (dblalock) [inline deeplab/transforms.py:28]:
```suggestion
           split (str): The dataset split to use; one of 'train', 'val', or 'test'. Default: ``'train```.
```

File: deeplab/transforms.py [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+           base_size (int): Initial size of the image and target before other augmentations. Default: ``512``.
+           min_resize_scale (float): The minimum value the samples can be rescaled. Default: ``0.5``.
+           max_resize_scale (float): The maximum value the samples can be rescaled. Default: ``2.0``.
```
Comment (dblalock) [inline deeplab/transforms.py:31]:
```suggestion
           min_resize_scale (float): The minimum value by which the samples can be rescaled. Default: ``0.5``.
           max_resize_scale (float): The maximum value by which the samples can be rescaled. Default: ``2.0``.
```

File: deeplab/transforms.py [hunk 9]
```diff
@@ -0,0 +1,271 @@
...
+            size=(final_size, final_size),
+            interpolation=TF.InterpolationMode.BILINEAR)
+        target_transforms = transforms.Resize(
+            size=(final_size, final_size),
+            interpolation=TF.InterpolationMode.NEAREST)
+    return both_transforms, image_transforms, target_transforms
+
+
+class RandomResizePair(torch.nn.Module):
+    """Resize the image and target to ``base_size`` scaled by a randomly sampled value.
```
Comment (dblalock) [inline deeplab/transforms.py:78]:
```suggestion
    """Resize the image and target to ``base_size`` times a randomly sampled value.
```

File: deeplab/transforms.py [hunk 10]
```diff
@@ -0,0 +1,271 @@
...
+    Args:
+        min_scale (float): the minimum value the samples can be rescaled.
+        max_scale (float): the maximum value the samples can be rescaled.
```
Comment (dblalock) [inline deeplab/transforms.py:81]:
```suggestion
        min_scale (float): the minimum value by which the samples can be rescaled.
        max_scale (float): the maximum value by which the samples can be rescaled.
```

File: deeplab/model.py [hunk 11]
```diff
@@ -0,0 +1,262 @@
...
+                             dice_weight: float = 0.0,
+                             init_fn: Optional[Callable] = None):
+    """Helper function to create a :class:`.ComposerClassifier` with a DeepLabv3(+) model. Logs
+        Mean Intersection over Union (MIoU) and Cross Entropy during training and validation.
+    From `Rethinking Atrous Convolution for Semantic Image Segmentation <https://arxiv.org/abs/1706.05587>`_
+        (Chen et al, 2017).
+    Args:
+        num_classes (int): Number of classes in the segmentation task.
+        backbone_arch (str, optional): The architecture to use for the backbone. Must be either
+            [``'resnet50'``, ``'resnet101'``]. Default: ``'resnet101'``.
```
Comment (dblalock) [inline deeplab/model.py:205]:
```suggestion
        backbone_arch (str, optional): The architecture to use for the backbone. Must be either
            ``'resnet50'`` or ``'resnet101'``. Default: ``'resnet101'``.
```

File: deeplab/model.py [hunk 12]
```diff
@@ -0,0 +1,262 @@
...
+
+    # Only apply initialization to classifier head if pre-trained weights are used
+    if init_fn:
+        if backbone_weights is None:
+            model.apply(init_fn)
+        else:
+            model.classifier.apply(init_fn)
+
+    if sync_bn and world_size > 1:
+        local_world_size = dist.get_local_world_size()
```
Comment (dblalock) [inline deeplab/model.py:166]: Do we need sync batchnorm? Just looking for ways to simplify what we give to users + customers.

File: deeplab/data.py [hunk 13]
```diff
@@ -0,0 +1,300 @@
...
+        self.both_transforms = both_transforms
+        self.image_transforms = image_transforms
+        self.target_transforms = target_transforms
+        split_to_dir = {'train': 'training', # map split names to ade20k image directories
+                        'val': 'validation',
+                        'test': 'test'}
+        self.split = split_to_dir[split]
+
+
+
```
Comment (dblalock) [inline deeplab/data.py:129]:
```suggestion
 
```

**Review by dblalock — APPROVED** [review]
This is awesome. I'm so excited to finally train a segmentation model.

But anyway, LGTM other than some local nitpicking. We'll also have to dedup the MCLI vs standalone configs, but that's a separate PR.

File: deeplab/model.py (hunk 12 above)
Comment (A-Jacobson) [reply deeplab/model.py:166]: it seems that we aren't using it but that's very much batch size dependent. It's likely going to be useful if someone has a smaller gpu or large enough image size to force their per device batch size down to ~4 or less.
